Feature/task/hchow/2308 student locations

hktest opened the pull request:
- Created page object structure for student locations tool.
- Added in tests to check that a user can add/edit user info successfully (build PO for a few additional pages for this check).
- Kept code as a clean as I can manage. 

Committed what was on my feature branch before my laptop gets taken away/swapped for new one -- that said, I had successfully run these tests on vagrant.  

Post commit review, it looks like I should add a "is_loaded" check on the view_table page object.
What's in there will work as is, but I will update for consistency post laptop swap along with any other PR feedback.

elliottyates commented:
@elliottyates

elliottyates commented on icommons_lti_tools/requirements/local.txt:
```diff
@@ -9,11 +9,12 @@ git+https://github.com/inducer/pudb.git#egg=pudb
+pep8==1.6.2
+pyvirtualdisplay==0.1.5
+selenium==2.48.0
+xlrd==0.9.4
```
Minor: can we order these alphabetically?

hktest replied:
Done, thanks!  
Also added the version number to the requirements that doesn't have a version.  @rascalking previously mentioned that it's a good practice.  
Also moved the requirement of  selenium-common==1.0 into this file (instead of base.txt) per Josie's feedback.

elliottyates commented on icommons_lti_tools/settings/local.py:
```diff
@@ -10,10 +10,11 @@
     'selenium_username': SECURE_SETTINGS.get('selenium_user'),
```
Minor: can we order these alphabetically?

hktest replied:
Fixed.

elliottyates commented on icommons_lti_tools/settings/local.py:
```diff
@@ -10,10 +10,11 @@
     'canvas_base_url': SECURE_SETTINGS.get('canvas_url'),
     'bulk_create_relative_url': SECURE_SETTINGS.get('bulk_create_relative_url'),
-
+    'student_locations_url': SECURE_SETTINGS.get('student_locations_url'),
```
Could we create a `'student_locations'` sub-dict like we do in other projects:
```
SELENIUM_CONFIG = {
...
    'student_locations': {
        ...
    }
}
```

hktest replied:
Fixed. Also updated the base_test_case to use this new sub-dict format.

elliottyates commented on selenium_tests/student_locations/page_objects/student_locations_edit_info_page_object.py:
```diff
@@ -0,0 +1 @@
        add_map_address.send_keys(address)
        # Click to save changes
        self.find_element(*Locators.submit_button).click()
        
```
If there's nothing in this file and we don't use it anywhere, let's just get omit it for now.

hktest replied:
Strange.  I was having laptop issues the day I check these in -  I don't know why the diff isn't showing here on github, but there is definitely content in this file that gets used by the tests.   Click on View to see the whole file.

hktest replied:
Fixed corrupted file.  Will need a quick PR on the code.

elliottyates commented on selenium_tests/student_locations/page_objects/student_locations_view_table_page_object.py:
```diff
@@ -0,0 +1 @@
            self.find_element(*Locators.email_address)
        except NoSuchElementException:
            return False
        return True
```
Ditto for unneeded files

hktest replied:
Ditto on comment above.  No idea why the diff isn't showing here, but click on "View" to see the whole file.  :(   This page object does get used by the test.

hktest replied:
Fixed corrupted file.  Will need a quick PR on the code.

elliottyates commented on selenium_tests/student_locations/student_locations_adding_user_tests.py:
```diff
@@ -0,0 +1 @@


if __name__ == "__main__":
    unittest.main()
```
Ditto for unneeded files

hktest replied:
Ditto above.  In addition to building out the page object, the bulk of the test logic is in this file.   Don't have my laptop but not sure if re-checking in these files will show the diff.

hktest replied:
Fixed corrupted file.  Will need a quick PR on the code.

elliottyates commented on selenium_tests/student_locations/student_locations_base_test_case.py:
```diff
@@ -0,0 +1,40 @@
+        else:
+            print '(User {} already logged in to PIN)'.format(cls.USERNAME)
+
+    @classmethod
```
No need for this if we're not adding anything to it. The `super()` bit is handled automatically if this method doesn't exist in the class, so this method can be removed entirely it looks like.

hktest replied:
Fixed.

elliottyates commented:
:+1: looks good as a base first test case (did the index page load). Doesn't look like there are any other tests at this point, but we can now add them incrementally.

hktest commented:
We are dealing with a corrupted branch --- the files look ok in pycharm and in sublime, but for whatever reasons, it's corrupted in sourcetree.  We'll get a clean copy of the files again and put it to a new branch.  Please hold on PR feedback on this one - apologies.  @Joy603 is going to help me with this piece.

hktest commented:
- Fixed corrupted files.  
- Updated per PR feedback.  
- Modified to follow the [latest] best practices/conventions.  
- Tested and ran successfully.  
- Needs PR as 3 files weren't PR due to the file corruption [now fixed]

elliottyates commented on icommons_lti_tools/requirements/base.txt:
```diff
@@ -1,18 +1,19 @@
-psycopg2==2.6.1
-ims_lti_py==0.6
-requests==2.7.0
-django-angular==0.7.13
```
`django-angular` and `django-crispy-forms` are missing in the updated version of this file -- are we sure that nothing in this project uses them? Have we run the full set of unit tests and selenium tests to ensure we haven't broken anything (given the coverage, we're probably still not sure we haven't broken anything, but that's better than nothing).

hktest replied:
- Great catch on django-crispy forms.  Fixed.  
- Hm. My local has the latest from develop, but django-angular wasn't on my local.txt copy to start with. Fixed and added.  Good point about running the full set of tests for this project -- my fault -- been so involved in this app that I only ran the tests for this app.  Will verify that regression suite runs fine (and unit tests - will check in with @Joy603 on how to do this) before merging to develop.  Thanks for the reminder here.

elliottyates commented on icommons_lti_tools/requirements/local.txt:
```diff
@@ -4,16 +4,17 @@
+selenium==2.48.0
+xlrd==0.9.4
+
+git+https://github.com/inducer/pudb.git#egg=pudb
```
We probably don't need `pudb` anymore, I don't think anyone's using that to debug.

hktest replied:
Fixed (removed).

elliottyates commented on icommons_lti_tools/settings/local.py:
```diff
@@ -7,12 +7,26 @@
+    'canvas_base_url': SECURE_SETTINGS.get('canvas_url'),
+
+    'student_loc_tool': {
+        'student_locations_url': 'courses/27/external_tools/39',
```
Recommendation for brevity, since `'student_loc_tool'` specifies the 'student location' part, this could probably be just `'url': ...`; for the same reason, if `testuser_` prefixes everything in the `test_loc` dict below, perhaps we can just get rid of it (and e.g. call the dict `test_user` or something?). It would make the lines of code we use to access these values shorter and more succinct. Just a thought / style suggestion -- not a requirement! (This is similar to the `Locator` classes -- since we put them inside a descriptive module, if we need to refer to them elsewhere we have enough information by referencing the `module`.`Locators` without a big long class name for the Locators.)

hktest replied:
Fixed, and it's a good style recommendation since some of the locators can get long (and I wasn't sure what the best way to do it since this is a new-ish format with using sub-dict). thanks!

elliottyates commented on selenium_tests/student_locations/page_objects/student_location_edit_info_page_object.py:
```diff
@@ -0,0 +1,50 @@
+This file models some of the page objects on the "Edit My Information" page of
+the Student Locations Tool
+"""
+from selenium.webdriver.common.by import By
```
Imports can also be alphabetized/organized

hktest replied:
Fixed.  Ordering helps fixes duplicate references too...I see what you did there.

elliottyates commented on selenium_tests/student_locations/page_objects/student_location_view_table_page_object.py:
```diff
@@ -0,0 +1,50 @@
+This file models some of the page objects on the "View Table" page of
+the Student Locations Tool
+"""
+from selenium.webdriver.common.by import By
```
ditto organization

hktest replied:
Fixed.

elliottyates commented on selenium_tests/student_locations/page_objects/student_location_view_table_page_object.py:
```diff
@@ -0,0 +1,50 @@
+    def is_loaded(self):
+        """ Determine if the view table page loaded by looking for specific
+        element on the page"""
+        self.focus_on_tool_frame()
```
An idea: these `is_loaded()` checks are all so similar that we could put most of this logic into selenium_common's BasePageObject in the future. We don't have to do this now, but maybe the next selenium tests we write we could give it a shot. It would reduce 'boilerplate code'.

hktest replied:
I won't make mods for this round here -- you're right that the logic is similar but the locator varies based on the page and we have is_loaded on every page object.  I'd like to pick your brains later on how to do this (I'm guessing... in selenium_common, you'd move the is_loaded method to it, and add a "find_locator" method that returns the locator which you'll pass in to the is_loaded method(?)  You then pass in the actual locator in the test itself by using the "find_locator" method?  Is that sort of the idea?  I'm not really strong enough in python yet to know how to best do this, so I'd love to work with someone to give this a shot for next round.

elliottyates commented on selenium_tests/student_locations/page_objects/student_location_view_table_page_object.py:
```diff
@@ -0,0 +1,50 @@
+        try:
+            element = self.find_element(*Locators.ADD_USER(testuser_email))
+        except NoSuchElementException:
+            return False
```
It is more common to return an object or `None`, or to return `True` or `False`, but not to mix the two. Could we return `None` here?

hktest replied:
Fixed.  (Did not know this!  thanks for pointer - will keep this in mind)

elliottyates commented on selenium_tests/student_locations/page_objects/student_location_view_table_page_object.py:
```diff
@@ -0,0 +1,50 @@
+        return element
+
+    def verify_successful_add(self, test_user_email):
+        """ determine if the user has been successfully added """
```
This is doing _almost_ the same thing as `add_user_element()`, so maybe it can read something like:
```
def verify_successful_add(self, test_user_email):
    return add_user_element(test_user_email) is not None
```

hktest replied:
Will check in offline since I could be missing something, but add_user_element() returns the element, and I need this to return True/False, because the test has an AssertTrue check...

elliottyates replied:
You may not need to do this; see [this comment](https://github.com/Harvard-University-iCommons/icommons_lti_tools/pull/392#discussion_r49814511).

elliottyates commented on selenium_tests/student_locations/page_objects/student_locations_index_page_object.py:
```diff
@@ -0,0 +1,40 @@
+
+
+class Locators(object):
+    """No page title associated with tool; checking by link text"""
```
This comment might be more appropriate in the `is_loaded()` method instead of here

hktest replied:
Fixed.

elliottyates commented on selenium_tests/student_locations/student_locations_add_user_tests.py:
```diff
@@ -0,0 +1,49 @@
+
+from selenium_tests.student_locations.student_locations_base_test_case \
+    import StudentLocationsBaseTestCase
+from selenium_tests.student_locations.page_objects.student_location_edit_info_page_object \
```
I notice here that some of the page objects are called `student_location_...` and others `student_locations_...` (plural) -- let's be consistent. Looks like the tool is plural, so let's stick with that.

hktest replied:
Fixed. Those were the renamed due to corrupted files.

elliottyates commented on selenium_tests/student_locations/student_locations_add_user_tests.py:
```diff
@@ -0,0 +1,49 @@
+class StudentLocationsEditInfoTests(StudentLocationsBaseTestCase):
+
+    def test_add_user(self):
+        """TLT-2308 - Add user info in student locations tool and
```
Nice

elliottyates commented on selenium_tests/student_locations/student_locations_add_user_tests.py:
```diff
@@ -0,0 +1,49 @@
+        org = test_data['testuser_org']
+        location = test_data['testuser_location']
+
+        # The "Edit my information" is not a relative link and must clicked
```
I'm not sure what 'relative link' means?

hktest replied:
D'oh. Re-worded.

elliottyates commented on selenium_tests/student_locations/student_locations_add_user_tests.py:
```diff
@@ -0,0 +1,49 @@
+        # The "Edit my information" is not a relative link and must clicked
+        # onto from the index page.
+        index_page = StudentLocationsIndexPageObject(self.driver)
+        index_page.click_on_edit_my_information_link()
```
I'm not sure what best practices are, but do we want to assert `is_loaded()` here?

hktest replied:
Fixed.

elliottyates commented on selenium_tests/student_locations/student_locations_add_user_tests.py:
```diff
@@ -0,0 +1,49 @@
+        edit_info.add_user_info(firstname, lastname, email, org, location)
+
+        # Go to View Table Page to verify user add
+        index_page.click_on_view_table_link()
```
ditto for asserting (that we're back on the index page)? Not sure if that's just a lot of overhead though.

hktest replied:
Fixed.  Seems a bit of an overhead since the focus here is now on view_table page, but i suppose it's a good idea to check the test workflow.

elliottyates commented on selenium_tests/student_locations/student_locations_add_user_tests.py:
```diff
@@ -0,0 +1,49 @@
+        view_table = StudentLocationsViewTablePageObject(self.driver)
+        self.assertTrue(view_table.is_loaded())
+        # Verify successful add of user
+        view_table.add_user_element(email)
```
Is this actually doing anything? Seems like the line below this is the only relevant one, in which case the comments on changing `verify_successful_add()` to call `add_user_element()` under the covers can be revised -- just remove the `add_user_element()` entirely and keep `verify_successful_add()` the way it is.

hktest replied:
Fixed and thanks! :)

elliottyates commented on selenium_tests/student_locations/student_locations_add_user_tests.py:
```diff
@@ -0,0 +1,49 @@
+        self.assertTrue(view_table.verify_successful_add(email))
+
+
+if __name__ == "__main__":
```
Can we remove these? You should be able to run test case files using:
```
python -m unittest ...
```
See the [documentation](https://docs.python.org/2/library/unittest.html#command-line-interface) for examples.

hktest replied:
Removed.  On my TODO to run it that way.

elliottyates commented on selenium_tests/student_locations/student_locations_index_page_tests.py:
```diff
@@ -0,0 +1,18 @@
+import unittest
+
+from selenium_tests.student_locations.student_locations_base_test_case \
```
alphabetical

hktest replied:
Fixed.

elliottyates commented on selenium_tests/student_locations/student_locations_index_page_tests.py:
```diff
@@ -0,0 +1,18 @@
+        self.assertTrue(studentloc_index_page.is_loaded())
+
+
+if __name__ == "__main__":
```
ditto we can remove this

hktest replied:
Fixed.

elliottyates commented:
:+1: with some suggestions for cleanup, and the provision that we ensure the libraries we're removing from base.txt are not being used. Nice work!

hktest commented:
Safe to merging - these selenium tests are specific to the student locations apps, and has no impact on other apps in the icommons_lti_tools project.
